File: backend/vibez/model_router.py
"""Shared task-based model routing for backend inference."""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import httpx


@dataclass(frozen=True)
class ModelRoute:
    provider: str
    model: str
    mode: str
    max_tokens: int
    temperature: float
    timeout_ms: int
    base_url: str | None = None
    dimensions: int | None = None


def default_manifest_path() -> Path:
    return Path("config/model-routing.json")


def _resolve_manifest_path(manifest_path: Path | str | None = None) -> Path:
    path = Path(manifest_path) if manifest_path is not None else default_manifest_path()
    if path.is_absolute():
        return path
    cwd_path = Path.cwd() / path
    if cwd_path.exists():
        return cwd_path
    return Path(__file__).resolve().parents[2] / path


def load_routes(manifest_path: Path | str | None = None) -> dict[str, ModelRoute]:
    path = _resolve_manifest_path(manifest_path)
    payload = json.loads(path.read_text())
    if payload.get("version") != 1:
        raise ValueError(f"unsupported model routing manifest version: {payload.get('version')}")
    routes_raw = payload.get("routes")
    if not isinstance(routes_raw, dict):
        raise ValueError("model routing manifest must include a routes object")

    routes: dict[str, ModelRoute] = {}
    for task_id, raw_route in routes_raw.items():
        if not isinstance(raw_route, dict):
            raise ValueError(f"route {task_id} must be an object")
        routes[task_id] = ModelRoute(
            provider=str(raw_route["provider"]),
            model=str(raw_route["model"]),
            mode=str(raw_route["mode"]),
            max_tokens=int(raw_route["max_tokens"]),
            temperature=float(raw_route["temperature"]),
            timeout_ms=int(raw_route["timeout_ms"]),
            base_url=(
                str(raw_route["base_url"])
                if raw_route.get("base_url") is not None
                else None
            ),
            dimensions=(
                int(raw_route["dimensions"])
                if raw_route.get("dimensions") is not None
                else None
            ),
        )
    return routes


def get_route(task_id: str, manifest_path: Path | str | None = None) -> ModelRoute:
    routes = load_routes(manifest_path)
    if task_id not in routes:
        raise KeyError(f"unknown model route: {task_id}")
    return routes[task_id]


def validate_route_requirements(manifest_path: Path | str | None = None) -> None:
    for route in load_routes(manifest_path).values():
        if route.provider == "openai" and not os.environ.get("OPENAI_API_KEY"):
            raise RuntimeError("OPENAI_API_KEY is required by model routing")
        if route.provider == "anthropic" and not os.environ.get("ANTHROPIC_API_KEY"):
            raise RuntimeError("ANTHROPIC_API_KEY is required by model routing")
        if route.provider == "ollama" and not (
            route.base_url or os.environ.get("OLLAMA_BASE_URL")
        ):
            raise RuntimeError("OLLAMA_BASE_URL is required by model routing")


def _build_messages(
    *,
    prompt: str | None,
    system: str | None,
    messages: list[dict[str, str]] | None,
) -> list[dict[str, str]]:
    if messages:
        return [
            {"role": str(message["role"]), "content": str(message["content"])}
            for message in messages
        ]
    if prompt is None:
        raise ValueError("prompt is required when messages are not supplied")
    payload: list[dict[str, str]] = []
    if system:
        payload.append({"role": "system", "content": system})
    payload.append({"role": "user", "content": prompt})
    return payload


def _usage_dict(usage: Any) -> dict[str, int]:
    if usage is None:
        return {"input_tokens": 0, "output_tokens": 0}
    return {
        "input_tokens": int(getattr(usage, "input_tokens", 0) or 0),
        "output_tokens": int(getattr(usage, "output_tokens", 0) or 0),
    }


def _parse_json_output(raw: str) -> Any:
    text = raw.strip()
    fenced = text.startswith("```") and text.endswith("```")
    if fenced:
        text = text.split("\n", 1)[1].rsplit("```", 1)[0].strip()
    return json.loads(text)


def _run_anthropic(
    route: ModelRoute,
    *,
    prompt: str | None,
    system: str | None,
    messages: list[dict[str, str]] | None,
) -> dict[str, Any]:
    import anthropic

    payload = _build_messages(prompt=prompt, system=None, messages=messages)
    with anthropic.Anthropic(api_key=os.environ.get("ANTHROPIC_API_KEY")) as client:
        response = client.messages.create(
            model=route.model,
            max_tokens=route.max_tokens,
            system=system,
            messages=payload,
        )
    text = "\n".join(
        block.text.strip()
        for block in getattr(response, "content", []) or []
        if getattr(block, "text", None)
    ).strip()
    return {
        "text": text,
        "usage": _usage_dict(getattr(response, "usage", None)),
    }


def _run_openai(
    route: ModelRoute,
    *,
    prompt: str | None,
    system: str | None,
    messages: list[dict[str, str]] | None,
) -> dict[str, Any]:
    from openai import OpenAI

    payload = _build_messages(prompt=prompt, system=system, messages=messages)
    client = OpenAI(
        api_key=os.environ.get("OPENAI_API_KEY"),
        timeout=max(route.timeout_ms / 1000, 1),
    )
    response = client.responses.create(
        model=route.model,
        input=payload,
        max_output_tokens=route.max_tokens,
        temperature=route.temperature,
    )
    return {
        "text": getattr(response, "output_text", "") or "",
        "usage": _usage_dict(getattr(response, "usage", None)),
    }


def _embed_openai(
    route: ModelRoute,
    *,
    texts: list[str],
    dimensions: int | None = None,
) -> list[list[float]]:
    from openai import OpenAI

    client = OpenAI(
        api_key=os.environ.get("OPENAI_API_KEY"),
        timeout=max(route.timeout_ms / 1000, 1),
    )
    kwargs: dict[str, Any] = {
        "model": route.model,
        "input": texts,
        "encoding_format": "float",
    }
    resolved_dimensions = dimensions or route.dimensions
    if resolved_dimensions:
        kwargs["dimensions"] = resolved_dimensions
    response = client.embeddings.create(**kwargs)
    return [list(item.embedding) for item in response.data]


def _run_ollama(
    route: ModelRoute,
    *,
    prompt: str | None,
    system: str | None,
    messages: list[dict[str, str]] | None,
) -> dict[str, Any]:
    payload = _build_messages(prompt=prompt, system=system, messages=messages)
    base_url = route.base_url or os.environ.get("OLLAMA_BASE_URL")
    if not base_url:
        raise RuntimeError("OLLAMA_BASE_URL is required by model routing")
    endpoint = f"{base_url.rstrip('/')}/api/chat"
    response = httpx.post(
        endpoint,
        json={
            "model": route.model,
            "messages": payload,
            "stream": False,
            "options": {
                "temperature": route.temperature,
                "num_predict": route.max_tokens,
            },
            "format": "json" if route.mode == "json" else None,
        },
        timeout=max(route.timeout_ms / 1000, 1),
    )
    response.raise_for_status()
    data = response.json()
    return {
        "text": data.get("message", {}).get("content", "") or "",
        "usage": {
            "input_tokens": int(data.get("prompt_eval_count", 0) or 0),
            "output_tokens": int(data.get("eval_count", 0) or 0),
        },
    }


def generate_text(
    task_id: str,
    *,
    prompt: str | None = None,
    system: str | None = None,
    messages: list[dict[str, str]] | None = None,
    manifest_path: Path | str | None = None,
) -> dict[str, Any]:
    route = get_route(task_id, manifest_path)
    if route.provider == "anthropic":
        result = _run_anthropic(route, prompt=prompt, system=system, messages=messages)
    elif route.provider == "openai":
        result = _run_openai(route, prompt=prompt, system=system, messages=messages)
    elif route.provider == "ollama":
        result = _run_ollama(route, prompt=prompt, system=system, messages=messages)
    else:
        raise ValueError(f"unsupported model provider: {route.provider}")
    return {
        **result,
        "provider": route.provider,
        "model": route.model,
    }


def generate_json(
    task_id: str,
    *,
    prompt: str | None = None,
    system: str | None = None,
    messages: list[dict[str, str]] | None = None,
    manifest_path: Path | str | None = None,
) -> dict[str, Any]:
    result = generate_text(
        task_id,
        prompt=prompt,
        system=system,
        messages=messages,
        manifest_path=manifest_path,
    )
    return {
        **result,
        "parsed": _parse_json_output(result["text"]),
    }


def embed_texts(
    task_id: str,
    texts: list[str],
    *,
    dimensions: int | None = None,
    manifest_path: Path | str | None = None,
) -> list[list[float]]:
    route = get_route(task_id, manifest_path)
    if route.mode != "embedding":
        raise ValueError(f"task {task_id} is not an embedding route")
    if route.provider == "openai":
        return _embed_openai(route, texts=texts, dimensions=dimensions)
    raise ValueError(f"unsupported embedding provider: {route.provider}")


def embed_text(
    task_id: str,
    text: str,
    *,
    dimensions: int | None = None,
    manifest_path: Path | str | None = None,
) -> list[float]:
    vectors = embed_texts(
        task_id,
        [text],
        dimensions=dimensions,
        manifest_path=manifest_path,
    )
    return vectors[0] if vectors else []

File: dashboard/src/lib/model-router.ts
import fs from "node:fs";
import path from "node:path";

import Anthropic from "@anthropic-ai/sdk";
import OpenAI from "openai";

export interface ModelRoute {
  provider: "openai" | "anthropic" | "openrouter" | "ollama";
  model: string;
  mode: "text" | "json" | "embedding";
  max_tokens: number;
  temperature: number;
  timeout_ms: number;
  base_url?: string;
  api_key_env?: string;
  dimensions?: number;
  context_window?: number;
}

type ModelRouteManifestEntry =
  Partial<ModelRoute> & {
    registry_route?: string;
  };

interface ModelMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

export interface ModelTextResult {
  text: string;
  model: string;
  provider: string;
  usage: {
    input_tokens: number;
    output_tokens: number;
  };
}

export interface ModelEmbeddingResult {
  vectors: number[][];
  model: string;
  provider: string;
}

const ROUTE_CACHE = new Map<string, Record<string, ModelRoute>>();
const REGISTRY_CACHE = new Map<string, Record<string, Record<string, string | number | boolean>>>();
const DEFAULT_OLLAMA_EMBEDDING_INPUT_MAX_CHARS = 1600;
const PROVIDER_CREDENTIAL_ENVS = {
  openai: ["VIBEZ_OPENAI_API_KEY", "OPENAI_API_KEY"],
  anthropic: ["VIBEZ_ANTHROPIC_API_KEY", "ANTHROPIC_API_KEY"],
  openrouter: ["VIBEZ_OPENROUTER_API_KEY", "OPENROUTER_API_KEY"],
} as const;

export function defaultManifestPath(): string {
  return process.env.VIBEZ_MODEL_ROUTING_PATH ||
    path.join(/* turbopackIgnore: true */ process.cwd(), "..", "config", "model-routing.json");
}

function defaultCognitionRegistryPath(): string {
  return process.env.VIBEZ_COGNITION_PRESETS_PATH ||
    process.env.PAIA_COGNITION_PRESETS_PATH ||
    "/Users/braydon/projects/experiments/paia-agent-runtime/config/cognition-presets.toml";
}

function resolveManifestPath(manifestPath = defaultManifestPath()): string {
  if (path.isAbsolute(manifestPath)) return manifestPath;
  const cwdCandidate = path.join(/* turbopackIgnore: true */ process.cwd(), manifestPath);
  if (fs.existsSync(cwdCandidate)) return cwdCandidate;
  return path.join(/* turbopackIgnore: true */ process.cwd(), "..", manifestPath);
}

function normalizeTomlTableName(rawName: string): string {
  return rawName.replace(/"([^"]+)"/g, "$1");
}

function parseTomlValue(rawValue: string): string | number | boolean {
  const withoutComment = rawValue.replace(/\s+#.*$/, "").trim();
  if (withoutComment.startsWith('"') && withoutComment.endsWith('"')) {
    return withoutComment.slice(1, -1);
  }
  if (withoutComment === "true") return true;
  if (withoutComment === "false") return false;
  const numeric = Number(withoutComment);
  if (Number.isFinite(numeric)) return numeric;
  return withoutComment;
}

function loadCognitionRegistry(
  registryPath = defaultCognitionRegistryPath(),
): Record<string, Record<string, string | number | boolean>> {
  const cached = REGISTRY_CACHE.get(registryPath);
  if (cached) return cached;
  if (!fs.existsSync(/* turbopackIgnore: true */ registryPath)) {
    throw new Error(`cognition registry not found: ${registryPath}`);
  }
  const tables: Record<string, Record<string, string | number | boolean>> = {};
  let currentTable = "";
  for (const line of fs.readFileSync(/* turbopackIgnore: true */ registryPath, "utf8").split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const tableMatch = trimmed.match(/^\[([^\]]+)\]$/);
    if (tableMatch) {
      currentTable = normalizeTomlTableName(tableMatch[1]);
      tables[currentTable] = tables[currentTable] || {};
      continue;
    }
    if (!currentTable) continue;
    const assignment = trimmed.match(/^([A-Za-z0-9_-]+)\s*=\s*(.+)$/);
    if (!assignment) continue;
    tables[currentTable][assignment[1]] = parseTomlValue(assignment[2]);
  }
  REGISTRY_CACHE.set(registryPath, tables);
  return tables;
}

function numberFromRegistry(value: string | number | boolean | undefined): number | undefined {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string") {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return parsed;
  }
  return undefined;
}

function stringFromRegistry(value: string | number | boolean | undefined): string | undefined {
  return typeof value === "string" && value ? value : undefined;
}

function routeModeFromRegistry(
  value: string | number | boolean | undefined,
): ModelRoute["mode"] | undefined {
  return value === "text" || value === "json" || value === "embedding" ? value : undefined;
}

function resolveRegistryCredentialEnv({
  provider,
  surface,
  registry,
}: {
  provider: ModelRoute["provider"];
  surface?: Record<string, string | number | boolean>;
  registry: Record<string, Record<string, string | number | boolean>>;
}): string | undefined {
  const assignment = registry['service_credential_assignments.vibez-monitor']?.[provider];
  const credentialAlias = stringFromRegistry(assignment);
  const credentialEnv = credentialAlias
    ? stringFromRegistry(registry[`credentials.${credentialAlias}`]?.env_var)
    : undefined;
  return credentialEnv || stringFromRegistry(surface?.api_key_env);
}

function resolveRegistryRoute(registryRouteId: string): Partial<ModelRoute> {
  const registry = loadCognitionRegistry();
  const route = registry[`model_routes.${registryRouteId}`];
  if (!route) {
    throw new Error(`unknown cognition registry route: ${registryRouteId}`);
  }
  const provider = stringFromRegistry(route.provider) as ModelRoute["provider"] | undefined;
  const surfaceName = stringFromRegistry(route.surface);
  const surface = surfaceName ? registry[`provider_surfaces.${surfaceName}`] : undefined;
  const timeoutSeconds =
    numberFromRegistry(route.request_timeout_seconds) ??
    numberFromRegistry(surface?.complete_timeout_seconds) ??
    numberFromRegistry(surface?.continue_timeout_seconds);

  return {
    provider,
    model: stringFromRegistry(route.model),
    mode: routeModeFromRegistry(route.mode),
    max_tokens: numberFromRegistry(route.max_tokens_default),
    temperature: numberFromRegistry(route.temperature_default),
    timeout_ms: timeoutSeconds ? timeoutSeconds * 1000 : undefined,
    base_url: stringFromRegistry(surface?.base_url),
    api_key_env: provider ? resolveRegistryCredentialEnv({ provider, surface, registry }) : undefined,
    context_window: numberFromRegistry(route.context_window),
  };
}

function finalizeRoute(routeId: string, entry: ModelRouteManifestEntry): ModelRoute {
  const registryRoute = entry.registry_route ? resolveRegistryRoute(entry.registry_route) : {};
  const merged = {
    ...registryRoute,
    ...entry,
  };
  delete merged.registry_route;
  if (!merged.provider || !merged.model || !merged.mode) {
    throw new Error(`incomplete model route: ${routeId}`);
  }
  return {
    provider: merged.provider,
    model: merged.model,
    mode: merged.mode,
    max_tokens: merged.max_tokens ?? 0,
    temperature: merged.temperature ?? 0,
    timeout_ms: merged.timeout_ms ?? 60_000,
    base_url: merged.base_url,
    api_key_env: merged.api_key_env,
    dimensions: merged.dimensions,
    context_window: merged.context_window,
  };
}

export function loadRoutes(manifestPath = defaultManifestPath()): Record<string, ModelRoute> {
  const resolvedPath = resolveManifestPath(manifestPath);
  const cached = ROUTE_CACHE.get(resolvedPath);
  if (cached) return cached;
  const payload = JSON.parse(fs.readFileSync(/* turbopackIgnore: true */ resolvedPath, "utf8")) as {
    version: number;
    routes: Record<string, ModelRouteManifestEntry>;
  };
  if (payload.version !== 1 || !payload.routes || typeof payload.routes !== "object") {
    throw new Error("invalid model routing manifest");
  }
  const routes = Object.fromEntries(
    Object.entries(payload.routes).map(([routeId, entry]) => [routeId, finalizeRoute(routeId, entry)]),
  );
  ROUTE_CACHE.set(resolvedPath, routes);
  return routes;
}

export function getRoute(
  taskId: string,
  routes: Record<string, ModelRoute> = loadRoutes(),
): ModelRoute {
  const route = routes[taskId];
  if (!route) {
    throw new Error(`unknown model route: ${taskId}`);
  }
  return route;
}

export function resolveProviderApiKey(
  provider: keyof typeof PROVIDER_CREDENTIAL_ENVS,
): string {
  for (const envName of PROVIDER_CREDENTIAL_ENVS[provider]) {
    const value = process.env[envName];
    if (value) return value;
  }
  return "";
}

function resolveRouteApiKey(route: ModelRoute): string {
  if (route.api_key_env) {
    return process.env[route.api_key_env] || "";
  }
  if (route.provider in PROVIDER_CREDENTIAL_ENVS) {
    return resolveProviderApiKey(route.provider as keyof typeof PROVIDER_CREDENTIAL_ENVS);
  }
  return "";
}

function validateRouteRequirements(route: ModelRoute): void {
  if (route.provider in PROVIDER_CREDENTIAL_ENVS) {
    const provider = route.provider as keyof typeof PROVIDER_CREDENTIAL_ENVS;
    if (!resolveRouteApiKey(route)) {
      throw new Error(`${route.api_key_env || PROVIDER_CREDENTIAL_ENVS[provider][0]} not configured`);
    }
  }
  if (route.provider === "ollama") {
    return;
  }
}

function normalizeEmbeddingDimensions(
  vector: number[],
  dimensions?: number,
): number[] {
  if (!dimensions || dimensions <= 0 || vector.length === dimensions) {
    return vector.map((value) => Number(value));
  }
  if (vector.length < dimensions) {
    throw new Error(
      `embedding dimension mismatch: requested ${dimensions}, got ${vector.length}`,
    );
  }
  const shortened = vector.slice(0, dimensions).map((value) => Number(value));
  const norm = Math.sqrt(
    shortened.reduce((sum, value) => sum + value * value, 0),
  );
  if (norm <= 0) return shortened;
  return shortened.map((value) => value / norm);
}

function ollamaEmbeddingInputMaxChars(): number {
  const value = Number.parseInt(process.env.VIBEZ_OLLAMA_EMBEDDING_INPUT_MAX_CHARS || "", 10);
  if (!Number.isFinite(value) || value <= 0) return DEFAULT_OLLAMA_EMBEDDING_INPUT_MAX_CHARS;
  return Math.max(256, Math.min(value, 8000));
}

function normalizeOllamaEmbeddingInputs(inputs: string[]): string[] {
  const maxChars = ollamaEmbeddingInputMaxChars();
  return inputs.map((input) => String(input || "").slice(0, maxChars));
}

function buildMessages({
  prompt,
  system,
  messages,
}: {
  prompt?: string;
  system?: string;
  messages?: ModelMessage[];
}): ModelMessage[] {
  if (messages && messages.length > 0) {
    return messages;
  }
  if (!prompt) {
    throw new Error("prompt is required when messages are not supplied");
  }
  const payload: ModelMessage[] = [];
  if (system) payload.push({ role: "system", content: system });
  payload.push({ role: "user", content: prompt });
  return payload;
}

function usageFromAnthropic(usage: Anthropic.Messages.Usage | undefined) {
  return {
    input_tokens: usage?.input_tokens ?? 0,
    output_tokens: usage?.output_tokens ?? 0,
  };
}

function usageFromOpenAI(response: { usage?: { input_tokens?: number; output_tokens?: number } }) {
  return {
    input_tokens: response.usage?.input_tokens ?? 0,
    output_tokens: response.usage?.output_tokens ?? 0,
  };
}

function usageFromOpenAIChat(response: {
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
    input_tokens?: number;
    output_tokens?: number;
  };
}) {
  return {
    input_tokens: response.usage?.prompt_tokens ?? response.usage?.input_tokens ?? 0,
    output_tokens: response.usage?.completion_tokens ?? response.usage?.output_tokens ?? 0,
  };
}

function parseJsonText(raw: string): unknown {
  const trimmed = raw.trim();
  const fenced = trimmed.match(/```(?:json)?\s*([\s\S]*?)```/i)?.[1]?.trim();
  return JSON.parse(fenced || trimmed);
}

async function runAnthropicRoute(
  route: ModelRoute,
  payload: ModelMessage[],
): Promise<ModelTextResult> {
  const system = payload
    .filter((message) => message.role === "system")
    .map((message) => message.content)
    .join("\n\n");
  const messages: Array<{ role: "user" | "assistant"; content: string }> = payload
    .filter((message) => message.role !== "system")
    .map((message) => ({
      role: message.role === "assistant" ? "assistant" : "user",
      content: message.content,
    }));
  const client = new Anthropic({ apiKey: resolveRouteApiKey(route) });
  const response = await client.messages.create({
    model: route.model,
    max_tokens: route.max_tokens,
    system: system || undefined,
    messages,
  });
  return {
    text: response.content
      .map((block) => ("text" in block && typeof block.text === "string" ? block.text : ""))
      .join("\n")
      .trim(),
    model: route.model,
    provider: route.provider,
    usage: usageFromAnthropic(response.usage),
  };
}

async function runOpenAIRoute(
  route: ModelRoute,
  payload: ModelMessage[],
): Promise<ModelTextResult> {
  const client = new OpenAI({
    apiKey: resolveRouteApiKey(route),
    timeout: route.timeout_ms,
  });
  const response = await client.responses.create({
    model: route.model,
    input: payload,
    max_output_tokens: route.max_tokens,
    temperature: route.temperature,
  });
  return {
    text: response.output_text || "",
    model: route.model,
    provider: route.provider,
    usage: usageFromOpenAI(response),
  };
}

async function runOpenRouterRoute(
  route: ModelRoute,
  payload: ModelMessage[],
): Promise<ModelTextResult> {
  if (!route.base_url) {
    throw new Error("openrouter base_url not configured");
  }
  const client = new OpenAI({
    apiKey: resolveRouteApiKey(route),
    baseURL: route.base_url,
    timeout: route.timeout_ms,
  });
  const response = await client.chat.completions.create({
    model: route.model,
    messages: payload,
    max_tokens: route.max_tokens,
    temperature: route.temperature,
    response_format: route.mode === "json" ? { type: "json_object" } : undefined,
  });
  return {
    text: response.choices[0]?.message?.content || "",
    model: route.model,
    provider: route.provider,
    usage: usageFromOpenAIChat(response),
  };
}

async function runOpenAIEmbeddingRoute(
  route: ModelRoute,
  inputs: string[],
  dimensions?: number,
): Promise<ModelEmbeddingResult> {
  const client = new OpenAI({
    apiKey: resolveRouteApiKey(route),
    timeout: route.timeout_ms,
  });
  const response = await client.embeddings.create({
    model: route.model,
    input: inputs,
    encoding_format: "float",
    dimensions: dimensions ?? route.dimensions,
  });
  return {
    vectors: response.data.map((item) =>
      normalizeEmbeddingDimensions(item.embedding as number[], dimensions ?? route.dimensions),
    ),
    model: route.model,
    provider: route.provider,
  };
}

async function runOpenRouterEmbeddingRoute(
  route: ModelRoute,
  inputs: string[],
  dimensions?: number,
): Promise<ModelEmbeddingResult> {
  if (!route.base_url) {
    throw new Error("openrouter base_url not configured");
  }
  const client = new OpenAI({
    apiKey: resolveRouteApiKey(route),
    baseURL: route.base_url,
    timeout: route.timeout_ms,
  });
  const response = await client.embeddings.create({
    model: route.model,
    input: inputs,
    encoding_format: "float",
  });
  return {
    vectors: response.data.map((item) =>
      normalizeEmbeddingDimensions(item.embedding as number[], dimensions ?? route.dimensions),
    ),
    model: route.model,
    provider: route.provider,
  };
}

async function runOllamaRoute(
  route: ModelRoute,
  payload: ModelMessage[],
): Promise<ModelTextResult> {
  const baseUrl = route.base_url || process.env.OLLAMA_BASE_URL || "http://localhost:11434";
  if (!baseUrl) throw new Error("OLLAMA_BASE_URL not configured");
  let response: Response;
  try {
    response = await fetch(`${baseUrl.replace(/\/$/, "")}/api/chat`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      signal: AbortSignal.timeout(route.timeout_ms),
      body: JSON.stringify({
        model: route.model,
        messages: payload,
        stream: false,
        options: {
          temperature: route.temperature,
          num_predict: route.max_tokens,
          num_ctx: route.context_window,
        },
        format: route.mode === "json" ? "json" : undefined,
      }),
    });
  } catch (error) {
    throw new Error(
      `ollama request failed for ${route.model}: ${
        error instanceof Error ? error.message : "request timed out"
      }`,
    );
  }
  if (!response.ok) {
    throw new Error(`ollama request failed: ${response.status}`);
  }
  const data = (await response.json()) as {
    message?: { content?: string };
    prompt_eval_count?: number;
    eval_count?: number;
  };
  return {
    text: data.message?.content || "",
    model: route.model,
    provider: route.provider,
    usage: {
      input_tokens: data.prompt_eval_count ?? 0,
      output_tokens: data.eval_count ?? 0,
    },
  };
}

async function runOllamaEmbeddingRoute(
  route: ModelRoute,
  inputs: string[],
  dimensions?: number,
): Promise<ModelEmbeddingResult> {
  const baseUrl = route.base_url || process.env.OLLAMA_BASE_URL || "http://localhost:11434";
  let response: Response;
  try {
    response = await fetch(`${baseUrl.replace(/\/$/, "")}/api/embed`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      signal: AbortSignal.timeout(route.timeout_ms),
      body: JSON.stringify({
        model: route.model,
        input: normalizeOllamaEmbeddingInputs(inputs),
        truncate: true,
      }),
    });
  } catch (error) {
    throw new Error(
      `ollama embed request failed for ${route.model}: ${
        error instanceof Error ? error.message : "request timed out"
      }`,
    );
  }
  if (!response.ok) {
    throw new Error(`ollama embed request failed: ${response.status}`);
  }
  const data = (await response.json()) as {
    embeddings?: number[][];
    embedding?: number[];
  };
  const rawEmbeddings = data.embeddings || (data.embedding ? [data.embedding] : []);
  return {
    vectors: rawEmbeddings.map((vector) =>
      normalizeEmbeddingDimensions(vector, dimensions ?? route.dimensions),
    ),
    model: route.model,
    provider: route.provider,
  };
}

export async function generateText({
  taskId,
  prompt,
  system,
  messages,
  manifestPath,
}: {
  taskId: string;
  prompt?: string;
  system?: string;
  messages?: ModelMessage[];
  manifestPath?: string;
}): Promise<ModelTextResult> {
  const routes = loadRoutes(manifestPath);
  const route = getRoute(taskId, routes);
  validateRouteRequirements(route);
  const payload = buildMessages({ prompt, system, messages });
  if (route.provider === "anthropic") {
    return runAnthropicRoute(route, payload);
  }
  if (route.provider === "openai") {
    return runOpenAIRoute(route, payload);
  }
  if (route.provider === "openrouter") {
    return runOpenRouterRoute(route, payload);
  }
  if (route.provider === "ollama") {
    return runOllamaRoute(route, payload);
  }
  throw new Error(`unsupported model provider: ${route.provider}`);
}

export async function generateJson<T>({
  taskId,
  prompt,
  system,
  messages,
  manifestPath,
}: {
  taskId: string;
  prompt?: string;
  system?: string;
  messages?: ModelMessage[];
  manifestPath?: string;
}): Promise<ModelTextResult & { parsed: T }> {
  const result = await generateText({ taskId, prompt, system, messages, manifestPath });
  return {
    ...result,
    parsed: parseJsonText(result.text) as T,
  };
}

export async function embedTexts({
  taskId,
  inputs,
  dimensions,
  manifestPath,
}: {
  taskId: string;
  inputs: string[];
  dimensions?: number;
  manifestPath?: string;
}): Promise<ModelEmbeddingResult> {
  const routes = loadRoutes(manifestPath);
  const route = getRoute(taskId, routes);
  validateRouteRequirements(route);
  if (route.mode !== "embedding") {
    throw new Error(`task ${taskId} is not an embedding route`);
  }
  if (route.provider === "openai") {
    return runOpenAIEmbeddingRoute(route, inputs, dimensions);
  }
  if (route.provider === "openrouter") {
    return runOpenRouterEmbeddingRoute(route, inputs, dimensions);
  }
  if (route.provider === "ollama") {
    return runOllamaEmbeddingRoute(route, inputs, dimensions);
  }
  throw new Error(`unsupported embedding provider: ${route.provider}`);
}

export async function embedText({
  taskId,
  input,
  dimensions,
  manifestPath,
}: {
  taskId: string;
  input: string;
  dimensions?: number;
  manifestPath?: string;
}): Promise<number[]> {
  const result = await embedTexts({
    taskId,
    inputs: [input],
    dimensions,
    manifestPath,
  });
  return result.vectors[0] || [];
}
